kili/queries/asset/helpers.py

```diff
@@ -0,0 +1,94 @@
+from tenacity import retry
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_random
+
+from kili.queries.asset.exceptions import MissingPropertyError
+
+
+def get_post_assets_call_process(
+    download_media: bool, local_media_dir: Optional[str], project_id: str
+):
+    """
+    Define the function to apply on assets after a paginated call of assets
+    Call the download_asset_media with the right parameters if download_media is True
+    Otherwise return assets without any post-processing
+    """
+    if download_media:
+        local_dir_path = (
+            Path(local_media_dir)
+            if local_media_dir is not None
+            else Path(f".cache/kili/projects/{project_id}/assets")
+        )
+
+        return lambda assets: download_asset_media(assets, local_dir_path)
+
+    return lambda assets: assets
+
+
+def get_file_extension_from_headers(url):
+    """guess the extension of a file with the url response headers"""
+    with requests.head(url, timeout=20) as header_response:
+        headers = header_response.headers
+        header_response.raise_for_status()
+        content_type = headers["content-type"]
+        return guess_extension(content_type)
+
+
+def get_download_path(url: str, external_id: str, local_dir_path: Path):
+    """Build the path to download a file the file in local."""
+    local_path = local_dir_path / external_id
+    extension = get_file_extension_from_headers(url)
+    if extension is not None:
+        local_path = local_path.with_suffix(extension)
+    return local_path
+
+
+@retry(stop=stop_after_attempt(2), wait=wait_random(min=1, max=2))
+def download_file(url: str, external_id: str, local_dir_path: Path):
+    """
+    Download a file by streming chunks of 1Mb
+    If the file already exists in local, it does not download it
+    """
+    local_path = get_download_path(url, external_id, local_dir_path)
+    if not local_path.is_file():
+        with requests.get(url, stream=True, timeout=20) as response:
+            response.raise_for_status()
+            with open(local_path, "wb") as file:
+                for chunk in response.iter_content(chunk_size=1024 * 1024):
+                    file.write(chunk)
+    return local_path
+
+
+def get_field_array(assets: List[dict], field: str):
+    """Get a list of an asset field from a generator of kili assets"""
+    field_array = []
+    for asset in assets:
+        value = asset.get(field)
+        if value is None:
+            raise MissingPropertyError(
+                f"The asset does not have the {field} property. Please add it to the fields when"
+                " querying assets "
+            )
+        field_array.append(value)
+    return field_array
+
+
+def download_asset_media(assets: List[dict], local_dir_path: Path) -> List[dict]:
+    """Download assets media in local"""
+    local_dir_path.mkdir(parents=True, exist_ok=True)
+    content_array: List[str] = get_field_array(assets, "content")
```

Comment:
Note on the perf, which is not mandatory: this will require to loop over the asset several times, which can be cumbersome with large asset list. You a can return a list of arrays, and unwrap it into content_array and content_id_array.

Reply:
and this could be generators too, to spare some memory in the same cases.

Reply:
yes right! I replaced it with generators. For the perf issue, I wait to see what we do for the json_content because it is a field that we might not query every time